sdk/identity/azure-identity/src/main/java/com/azure/identity/ManagedIdentityCredential.java

```diff
@@ -83,13 +99,51 @@ public final class ManagedIdentityCredential implements TokenCredential {
-            managedIdentityServiceCredential = new AksExchangeTokenCredential(clientIdentifier, clientBuilder.build());
+            managedIdentityServiceCredential = new AksExchangeTokenCredential(clientIdentifier, clientBuilder
+                .identityClientOptions(updateIdentityClientOptions(ManagedIdentityType.AKS,
+                    identityClientOptions, configuration))
+                .build());
         } else {
-            managedIdentityServiceCredential = new VirtualMachineMsiCredential(clientId, clientBuilder.build());
+            managedIdentityServiceCredential = new VirtualMachineMsiCredential(clientId, clientBuilder
+                .identityClientOptions(updateIdentityClientOptions(ManagedIdentityType.VM,
+                    identityClientOptions, configuration))
+                .build());
         }
         LoggingUtil.logAvailableEnvironmentVariables(LOGGER, configuration);
     }
 
+    private IdentityClientOptions updateIdentityClientOptions(ManagedIdentityType managedIdentityType,
+                                             IdentityClientOptions clientOptions, Configuration configuration) {
+        switch (managedIdentityType) {
+            case APP_SERVICE:
+                return clientOptions
+                    .setManagedIdentityType(ManagedIdentityType.APP_SERVICE)
+                    .setManagedIdentityParameters(new ManagedIdentityParameters()
+                        .setMsiEndpoint(configuration.get(Configuration.PROPERTY_MSI_ENDPOINT))
+                        .setMsiSecret(configuration.get(Configuration.PROPERTY_MSI_SECRET))
+                        .setIdentityEndpoint(configuration.get(Configuration.PROPERTY_IDENTITY_ENDPOINT))
+                        .setIdentityHeader(configuration.get(Configuration.PROPERTY_IDENTITY_HEADER)));
+            case SERVICE_FABRIC:
+                return clientOptions
+                    .setManagedIdentityType(ManagedIdentityType.SERVICE_FABRIC)
+                    .setManagedIdentityParameters(new ManagedIdentityParameters()
+                        .setIdentityServerThumbprint(configuration.get(PROPERTY_IDENTITY_SERVER_THUMBPRINT))
+                        .setIdentityEndpoint(configuration.get(Configuration.PROPERTY_IDENTITY_ENDPOINT))
+                        .setIdentityHeader(configuration.get(Configuration.PROPERTY_IDENTITY_HEADER)));
+            case ARC:
+                return clientOptions
+                    .setManagedIdentityType(ManagedIdentityType.ARC)
+                    .setManagedIdentityParameters(new ManagedIdentityParameters()
+                        .setIdentityEndpoint(configuration.get(Configuration.PROPERTY_IDENTITY_ENDPOINT)));
+            case VM:
+                return clientOptions.setManagedIdentityType(ManagedIdentityType.VM);
+            case AKS:
+                return clientOptions.setManagedIdentityType(ManagedIdentityType.AKS);
+            default:
```

Comment:
is `NONE` really a valid value? should this be an error?

Reply:
Removed NONE

Reply:
Hmm. I'm not sure that's entirely right either - now won't the default value be `VM`? We probably need a `NONE` or `DEFAULT` to indicate the unchosen value?

Reply:
yeah in our code flow we default to VM, that's our default based on the logic we follow.
None or Unchosen are invalid states to have and are equivalent to the value being null.